ouroboros-consensus/test-consensus/Test/Consensus/Mempool.hs

```diff
@@ -122,6 +130,191 @@ prop_Mempool_InvalidTxsNeverAdded setup =
+      } = mcts
+
+    runCapacityTest ::  forall m.
+                      ( MonadAsync  m
+                      , MonadMask   m
+                      , MonadFork   m
+                      )
+                    => [GenTx TestBlock]
+                    -> TestMempool m
+                    -> m Property
+    runCapacityTest txs testMempool@TestMempool{getTraceEvents} =
+      withRegistry $ \registry -> do
+        env@MempoolCapTestEnv
+          { mctEnvAddedTxs
+          , mctEnvRemovedTxs
+          } <- initMempoolCapTestEnv
+        void $ forkAddValidTxs  env registry testMempool txs
+        void $ forkUpdateLedger env registry testMempool txs
+
+        -- Before we check the order of events, we must block until we've:
+        -- * Added all of the transactions to the mempool
+        -- * Removed all of the transactions from the mempool
+        atomically $ do
+          envAdded   <- readTVar mctEnvAddedTxs
+          envRemoved <- readTVar mctEnvRemovedTxs
+          check $  envAdded   == fromIntegral (length txs)
+                && envRemoved == fromIntegral (length txs)
+
+        -- Check the order of events
+        events <- getTraceEvents
+        pure $ checkTraceEvents events
+
+    -- | Spawn a new thread which continuously attempts to fill the mempool to
+    -- capacity until no more transactions remain. This should block whenever
+    -- attempting to add more transactions while the mempool is full.
+    forkAddValidTxs ::   forall m.
+                       ( MonadAsync m
+                       , MonadFork m
+                       , MonadMask m
+                       , MonadSTM m
+                       )
+                    => MempoolCapTestEnv m
+                    -> ResourceRegistry m
+                    -> TestMempool m
+                    -> [GenTx TestBlock]
+                    -> m (Thread m ())
+    forkAddValidTxs env registry testMempool txs =
+      forkThread registry $ addValidTxs env testMempool txs
+
+    -- | Recursively attempt to fill the mempool to capacity until no further
+    -- transactions remain. This should block whenever attempting to add more
+    -- transactions while the mempool is full.
+    addValidTxs :: MonadSTM m
+                => MempoolCapTestEnv m
+                -> TestMempool m
+                -> [GenTx TestBlock]
+                -> m ()
+    addValidTxs env testMempool txs = case txs of
+      [] -> pure ()
+      ts -> do
+        let TestMempool{mempool} = testMempool
+            Mempool{addTxs}      = mempool
+
+        -- Attempt to fill the mempool to capacity
+        let MempoolCapacity mpCap    = mctsCapacity
+            (txsToAdd, txsRemaining) = ( take (fromIntegral mpCap) ts
+                                       , drop (fromIntegral mpCap) ts
+                                       )
+        _ <- addTxs txsToAdd
+        atomically $ incrementEnvAddedTxs env (fromIntegral $ length txsToAdd)
+
+        addValidTxs env testMempool txsRemaining
```

Comment:
Hmm, I added the following lines to do some tracing (the right thing to do would be *label*ling):
```haskell
curSize <- atomically $ length . snapshotTxs <$> getSnapshot mempool
traceM $ "Adding " <> show (length txsToAdd) <>
         " with room for " <> show (mpCap - fromIntegral curSize)
```
And I get: `Adding X with room for Y` where either `X == Y`, `Y == 0` or `X < Y`. I would also like to see `X > Y > 0`, but that doesn't happen. For example: adding 5 transactions while there is only room for 3, which would block after the 3rd transaction. Currently, the only time it blocks in the test is for the first transaction (`Y == 0`), never for the *n*th transaction.

Can you tweak the test so that `X > Y > 0` occurs? Adding a label for this would be nice too.

Reply:
As discussed on Slack with @mrBliss, I think that the current test already covers this. 

So the test never tries to add `X with room for Y` where `X > Y > 0`. However, as it stands, the test tries to add `X with room for Y` where `Y == 0` and then, immediately after that, tries to fill the `Mempool` again (tries to add `X with room for Y` where `Y == 0`) which should be a blocking operation in the same way that `X > Y > 0` should be. 

Here's how the test works:

The test generates some transactions (greater than the generated capacity) and spawns a thread that recursively tries to fill the `Mempool` to its capacity. The logic in `implAddTxs.validateNew` will cause this thread to block when the `Mempool` is already at capacity. This follows a very similar same code path as trying to add `X + n` transactions in a single call to `implAddTxs` where `X + n > Y > 0`. Essentially, `addTxs (a ++ b)` follows the same logic as `addTxs a ; addTxs b`. This is because `implAddTxs` actually calls upon itself with the remaining transactions that couldn't be added to the mempool. So filling it to capacity and then trying to add more in separate `implAddTxs` calls is essentially the same thing.

There is also another thread that is spawned which recursively attempts to remove transactions from the `Mempool` and add them to the ledger. It does this until all of the generated transactions have been added to the ledger (`envRemoved == fromIntegral (length txs)`).

Reply:
Actually, there could be an issue in which the thread that removes txs from the `Mempool` manages to remove them before we attempt to add more txs. If this happens, we wouldn't block.